### PR TITLE
Render videos while building documentation

### PR DESCRIPTION
**Pull request by behackl**
## List of Changes
- Build a Sphinx extension that creates a `.. manim::` directive so that videos are generated directly when building the documentation.

## Motivation
I don't really like that gifs + videos have to be added to the source files of the documentation. 

## Explanation for Changes
With this, the figures are generated from the code in the documentation.

## Testing Status
Building the documentation works locally.

## Acknowledgement
- [x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

### CODE REVIEW

**Comment by behackl** [comment]
Okay, so this is something like a proof of concept for a `.. manim::` Sphinx directive that allows rendering videos directly when building the documentation.

Before I start polishing / documenting this, I'd be interested in opinions. :-)

File: docs/source/conf.py [hunk 1]
```diff
@@ -29,11 +29,11 @@
 # ones.
 extensions = [
     "sphinx.ext.autodoc",
-    "sphinx_rtd_theme",
     "recommonmark",
     "sphinx.ext.napoleon",
     "sphinx.ext.autosummary",
     "sphinx.ext.doctest",
+    "manim.sphinxext.manim_directive",
```
Comment (leotrs) [inline docs/source/conf.py:36]: Question: I've never defined a sphinx extension so I don't know what is the standard here. Do they usually go within the main package itself? In the current way, the end-user will be able to access the directive which is not terribly wrong but seems unnecessary. Any way we can define the directive within the docs/ folder instead? This folder is not shipped to the final user.

File: manim/sphinxext/manim_directive.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+
+TEMPLATE = r"""
+{% if include_source %}
+{{ source_block }}
+{% endif %}
+
+.. raw:: html
+
+    <video style="width: 100%;" controls src="./{{ clsname }}.mp4"></video>
```
Comment (leotrs) [inline manim/sphinxext/manim_directive.py:80]: We should allow for including .gif and .png files as well. See [here](https://github.com/ManimCommunity/manim/pull/394/files#diff-87c85d2172f54987338d92c91d885088R101)

File: manim/sphinxext/manim_directive.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+
+def setup(app):
+    import manim
+    setup.app = app
+    setup.config = app.config
+    setup.confdir = app.confdir
+    app.add_directive('manim', ManimDirective)
+
+    metadata = {'parallel_read_safe': True, 'parallel_write_safe': True}
```
Comment (leotrs) [inline manim/sphinxext/manim_directive.py:69]: Question: what exactly do these do? There has been some discussion about moving manim to a multiprocessing setup to render scenes in parallel, so I just want to know how this will interact with that.

File: manim/sphinxext/manim_directive.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        source_block = '\n'.join(source_block)
+
+        rendered_template = jinja2.Template(TEMPLATE).render(
+            include_source=include_source,
+            source_block=source_block,
+            clsname=clsname
+        )
+        state_machine.insert_input(rendered_template.split('\n'), 
+                                   source=document.attributes['source'])
+        exec('\n'.join(self.content) + f'\n\n{clsname}()', globals())
```
Comment (leotrs) [inline manim/sphinxext/manim_directive.py:54]: Question: I'm a bit confused as to when in the process manim is actually being executed. I think it's this line, no? Could you expand upon this a bit? It seems to be directly executing what's in the content of the code block. Does this mean that every scene example will have to include `from manim import *` at the top, for example?

**Comment by leotrs** [comment]
Another great suggestion, thanks @behackl.

I see two main issues to sort out here.

First, this means that building the docs will take a long time (manim is not the most efficient software eh). Especially after doing `make clean` -- this will basically mean that all scenes in the whole documentation will need to be rebuilt.

Second, there has been ample discussion about _not_ storing rendered manim scenes in the documentation (or indeed in this repository at all). See #308 and discussion therein. If we implemented either a different repository, or a custom CDN for this, would this directive be able to use those?

File: docs/source/conf.py (hunk 1 above)
Comment (behackl) [reply docs/source/conf.py:36]: I think we are pretty flexible where this should go in the end. For this proof-of-concept, I more or less imitated what `matplotlib` does with its `.. plot::` directive, they have it in the package itself.

I don't really mind either way -- but it might be better to move it as currently it makes the doctests fail (didn't investigate in detail yet).

File: manim/sphinxext/manim_directive.py (hunk 3 above)
Comment (behackl) [reply manim/sphinxext/manim_directive.py:69]: See https://www.sphinx-doc.org/en/master/extdev/index.html -- it does not refer to whatever the extension executes, but rather to reading the documentation source and writing the html files.

File: docs/source/conf.py (hunk 1 above)
Comment (naveen521kk) [reply docs/source/conf.py:36]: I prefer having it as a new package (maybe a new repo) and not inside the main module manim. Because users never require this.

File: manim/sphinxext/manim_directive.py (hunk 4 above)
Comment (behackl) [reply manim/sphinxext/manim_directive.py:54]: Yes, you are right. This line takes the code given to the extension, together with the provided class name, and first executes the block and then calls `clsname()`. This triggers building the documentation.

(Currently, this is a bit (huge understatement) rough, as there are no customization options _at all_. This is the main point I'd like to still improve, before this PR is actually ready.)

File: docs/source/conf.py (hunk 1 above)
Comment (behackl) [reply docs/source/conf.py:36]: Let me say again: this is just a proof of concept, and not my final suggestion. I'm happy to move it to the documentation. :-)

**Comment by behackl** [comment]
> First, this means that building the docs will take a long time (manim is not the most efficient software eh). Especially after doing `make clean` -- this will basically mean that all scenes in the whole documentation will need to be rebuilt.

True. I don't think this is a big issue though: building the documentation is, AFAIK, not included in the pipeline -- and I also think that there could be a flag that skips compilation of the videos, if rendering `reST` should become a part of the pipeline as well.

And a long building time of the documentation should not be too troublesome, in the end. (Maybe depending on how the documentation is deployed currently.)
 
> Second, there has been ample discussion about _not_ storing rendered manim scenes in the documentation (or indeed in this repository at all). See #308 and discussion therein. If we implemented either a different repository, or a custom CDN for this, would this directive be able to use those? Tagging in @kolibril13 (who is AFK for a week at the moment)

I saw the discussion. This is sort of a contribution to it, in the sense that the video files / gifs / pictures do not really have to be stored in the repository, but only generated when deploying. At least if deploying the documentation does not require a pushed git branch with all the stuff in it... which might be the case currently, now that I think about it. ;-)

(I might be able to help find a way around that, though.)

Anyways, the directive would still be helpful: it enables us to keep the code for creating the animations neatly organized within the documentation itself. If you decide that the videos should be provided via a CDN, they can be put there (automatically) after generating them from the documentation.

**Comment by leotrs** [comment]
> the video files / gifs / pictures do not really have to be stored in the repository, but only generated when deploying.

Oh snap I think this makes much more sense! So we wouldn't need a CDN or repository at all, we would just use ReadTheDocs as our hosting.... 

I think we would still have to iron a few things, but I like this a lot, if we can pull it off. @eulertour @naveen521kk @kolibril13

**Comment by naveen521kk** [comment]
> ReadTheDocs as our hosting....

Do you think we can install manim on little CI readthedocs gives? If we can some how then nice.
Or considerably, we can move to GitHub pages or Gitlab pages.

**Comment by leotrs** [comment]
Building the docs already requires manim to be installed, so that should be no problem, as it's already being done.

The main problem would be hosting size I think.

**Comment by naveen521kk** [comment]
image
I see this in https://docs.readthedocs.io/en/stable/choosing-a-site.html
>  Not sure how much space RTD gives us.

I also suspect that. But I can see Jsdelivr saying a yes that they will provide a CDN for use which is nice also we can use storage bucket ibm provides and is done.

**Comment by behackl** [comment]
https://docs.readthedocs.io/en/stable/builds.html - here are some more concrete numbers, but I do not find any information about their space limitation. I don't think we will hit the limit too soon, though.

Thanks for the feedback so far, by the way! I think I will continue to bring this branch to a more stable form. (And, at least on this PR, without a particular CDN in mind. This can be taken care of separately; first building everything has to work out properly.)

**Comment by behackl** [comment]
We are getting somewhere. There are still some hacks here and there, and documentation is also not really existent, but here is an illustration of what is currently possible:

The code

```reST

Autogenerating Videos in Documentation
======================================

.. manim:: TestExample
    :display_source:

    from manim import *

    class TestExample(Scene):
        def construct(self):
            dot = Dot()
            self.add(dot)
            self.wait(1)


.. manim:: TestExample2
    :display_source:
    :save_as_gif:

    class TestExample2(Scene):
        def construct(self):
            dot = Dot()
            self.play(ShowCreation(dot))
            self.wait(1)


.. manim:: TestExample3
    :save_last_frame:

    class TestExample3(Scene):
        def construct(self):
            dot = Dot()
            self.add(dot)
            self.wait(1)
            dot2 = Dot()
            dot2.next_to(dot, RIGHT)
            self.play(ShowCreation(dot2))
            self.wait(1)

```
renders into the page hosted at https://manim.hackl.space/examples/playing_around.html.

TL;DR: The generated videos / gifs / pictures can be customized quite a bit by options like `:display_source:` and `:save_as_gif` etc.

Question: what should actually be customizable? Should custom resolutions in the documentation also be possible? Or do we want to stick to the / some default?

**Comment by behackl** [comment]
> I see two options here: either we implement a few of the necessary config options, OR we find a way to add some hidden code at the top of each code block. 

If we implement something that is able to recognize and then hide the corresponding lines, we'd have to be careful I suppose. Always removing everything before `class ...` seems a little brutal, and parsing line by line might be difficult if some arguments are passed in multiple lines.

> If so, then we won't need to implement dozens of `:sphinx_options:`.

I suppose that with `:display_source:` and `:save_last_frame:` the most important options for day-to-day documentation writing business are more or less covered.

In theory, it would also be possible to implement something that is able to customize everything by doing something like

```reST

.. manim::
    :render_config: {'some_key': some_value, ...}
    
    class TestExample(Scene):
        ...
```

I think I'd prefer an approach like this over having to parse code in order to decide what should be hidden. (Also because I think that there should not be too many figures where we would want to customize *a lot* of options here.) :-)

**Comment by leotrs** [comment]
That makes a lot more sense actually. So let's try to see how far we can go with `:display_source:` and `:save_last_frame:` and perhaps `:low_quality:` et al?

In the odd chance that a scene needs to change the config, we might as well show the code that does that.

**Comment by behackl** [comment]
- b9f2974d adds a custom HTML formatting for videos with included source code that looks like this:
  image

- cbabe7d7 allows using the directive in doctests settings (i.e., where lines are prepended by `>>> ` and `... `),
- ef5185be allows choosing custom render quality via the `:quality:` option
- c8b34e01 adds some documentation.

I think this covers basic functionality more or less.

File: docs/source/_static/custom.css [hunk 5]
```diff
@@ -7,3 +7,20 @@ img.align-center {
...
+}
+
+.manim-example pre {
+    border-left: none;
+    margin: 0;
+    padding: 0 0 14px 0;
+}
+
+.manim-example {
+    border-left: 2px solid #eee;
```
Comment (naveen521kk) [inline docs/source/_static/custom.css:22]: Suggestion: 2px would be very small and can be increased. Also, let the border be for all sides/.

File: docs/source/manim_directive.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+        ]
+        source_block = "\n".join(source_block)
+
+        file_writer_config_code = [
+            f'config["frame_rate"] = {frame_rate}',
+            f'config["pixel_height"] = {pixel_height}',
+            f'config["pixel_width"] = {pixel_width}',
+            'file_writer_config["media_dir"] = "./source/media"',
+            'file_writer_config["images_dir"] = "./source/media/images"',
+            'file_writer_config["video_dir"] = "./source/media/videos"',
```
Comment (naveen521kk) [inline docs/source/manim_directive.py:155]: Why not use `os` module path function?
like `os.path.join('source','media')`?

File: docs/source/manim_directive.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+            *file_writer_config_code,
+            *user_code,
+            f"{clsname}()",
+        ]
+        exec("\n".join(code), globals())
+
+        # copy video file to output directory
+        if not (save_as_gif or save_last_frame):
+            filename = f"{clsname}.mp4"
+            filesrc = f"source/media/videos/{qualitydir}/{filename}"
```
Comment (naveen521kk) [inline docs/source/manim_directive.py:177]: Why not os modules path function here?

File: docs/source/manim_directive.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+            destfile = os.path.join(dest_dir, filename)
+            shutil.copyfile(filesrc, destfile)
+        elif save_as_gif:
+            filename = f"{clsname}.gif"
+            filesrc = f"source/media/videos/{qualitydir}/{filename}"
```
Comment (naveen521kk) [inline docs/source/manim_directive.py:182]: Why not os modules path function here?

File: docs/source/manim_directive.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        elif save_last_frame:
+            filename = f"{clsname}.png"
+            filesrc = f"source/media/images/{clsname}.png"
```
Comment (naveen521kk) [inline docs/source/manim_directive.py:185]: Why not os modules path function here?

File: docs/Makefile [hunk 10]
```diff
@@ -14,10 +14,11 @@ help:
...
 .PHONY: help Makefile
 
-# Like make clean but also remove files generated by autosummary
-cleanall:
-	make clean
+# Like make clean but also remove files generated by autosummary and
+# rendered videos.
+cleanall: clean
 	rm source/reference/*
+	rm -rf source/media
```
Comment (naveen521kk) [inline docs/Makefile:21]: Why not make the same change in `make.bat`?

**Review by naveen521kk — CHANGES_REQUESTED** [review]
I found something that could possibly break testing in windows. I seriously think that that we need to have a test which checks whether the docs works in github actions.

File: docs/source/_static/custom.css (hunk 5 above)
Comment (behackl) [reply docs/source/_static/custom.css:22]: No. This should be consistent with the current appearance of source code blocks in the documentation.

File: docs/source/manim_directive.py (hunk 6 above)
Comment (behackl) [reply docs/source/manim_directive.py:155]: Good catch on all of these! I'll take care of it.

**Comment by behackl** [comment]
I've replaced the statically included source files in 0a11b2fd, and replaced them with the manim directive.

The resulting documentation can be inspected here: https://manim.hackl.space/tutorials/quickstart.html + https://manim.hackl.space/tutorials/a_deeper_look.html

File: docs/Makefile (hunk 10 above)
Comment (behackl) [reply docs/Makefile:21]: I don't think that the `make.bat` even has a `cleanall` directive, as of right now. I am not a Windows person, if you know how to write an appropriate equivalent, please make a suggestion.

File: docs/source/manim_directive.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+
+    def run(self):
+        clsname = self.arguments[0]
+
+        display_source = "display_source" in self.options
+        save_as_gif = "save_as_gif" in self.options
+        save_last_frame = "save_last_frame" in self.options
+        assert not (save_as_gif and save_last_frame)
+
+        frame_rate = 60
```
[the remark this answers is not in the document]
Comment (leotrs) [reply docs/source/manim_directive.py:98]: `config['frame_rate']`

File: docs/source/manim_directive.py [hunk 12]
```diff
@@ -53,29 +140,38 @@ def run(self):
...
+            *user_code,
             f'{clsname}()'
         ]
         exec('\n'.join(code), globals())
 
         # copy video file to output directory
         if not (save_as_gif or save_last_frame):
             filename = f'{clsname}.mp4'
-            filesrc = f'source/media/videos/1080p60/{filename}'
+            filesrc = f'source/media/videos/{qualitydir}/{filename}'
```
Comment (leotrs) [inline docs/source/manim_directive.py]: Comment: once #337 is closed, this will have to change. Let's make a note to come back to this

File: docs/Makefile (hunk 10 above)
Comment (leotrs) [reply docs/Makefile:21]: I added `cleanall` in the Makefile, and should have changed make.bat as well

File: docs/source/manim_directive.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (behackl) [reply docs/source/manim_directive.py]: Thanks! e1139f2e

File: docs/source/manim_directive.py (hunk 6 above)
Comment (behackl) [reply docs/source/manim_directive.py]: Taken care of via 2dc15e84.

File: docs/Makefile [hunk 13]
```diff
@@ -18,6 +18,7 @@ help:
 cleanall:
 	make clean
 	rm source/reference/*
+	rm -rf media
```
[the remark this answers is not in the document]
Comment (behackl) [reply docs/Makefile]: The current version of the makefile fully clears `source/reference` for me (and I didn't change anything there). Weird.

**Comment by behackl** [comment]
I just realized another positive aspect for the documentation: the videos / images included there are then automatically rerendered with the latest version of the engine.

Anyways, this branch is feature complete in the sense that it introduces basic functionality of the directive. Review at will. :-)

**Comment by leotrs** [comment]
Agreed! Before merging, I'm interested in what @kolibril13 has to say regarding this addition and the secondary repository for examples. If we merge this PR, it seems like we can forget about a second repo for a while.

**Comment by kolibril13** [comment]
Wow, this is really really nice!

I also like the idea of using:

```python
    config['save_last_frame'] = True
    class MyScene2(Scene):
        def construct(self):
            some_code_here()
```
and then hiding the first lines, because then one can write the manim scripts and just copy them into the docs to be rendered.
Maybe one could write a regular expression for that, so it won't be necessary to put this in the exact line.
The regular expression could be something like:
Look in the first 10 lines for "save_last_frame" and check if it is followed by True.

But the other idea, to use the sphinx internal syntax would be also ok.

I agree that in this case, a second repo would make no sense. 
And further, it makes is so much more convenient to quickly add new scenes without carrying about any references to further clouds.
Is there maybe an option that rebuilds care only the changed files in the docs, and not everything?
Then the building process could be fast.
And the whole build of everything could e.g. be automated to happen once in a week.

**Review by kolibril13 — APPROVED** [review]

**Comment by behackl** [comment]
> I also like the idea of using:
> 
> ```python
>     config['save_last_frame'] = True
>     class MyScene2(Scene):
>         def construct(self):
>             some_code_here()
> ```
> 
> and then hiding the first lines, because then one can write the manim scripts and just copy them into the docs to be rendered.
> [...]
> But the other idea, to use the sphinx internal syntax would be also ok.

When implementing more flexibility to the directive in a follow-up PR, I think I'd still prefer to allow passing a dict in the directive options over parsing and (optionally) hiding written lines of code.

Imagine, otherwise, that you would like to hide several config options (like in order to only save the last frame or so), but at the same time show how the background color can be changed. This is why I think it makes sense to have the option to show how to set some options (by putting them in the code block) while still hiding others (by passing them as options to the directive).

> I agree that in this case, a second repo would make no sense.

I fully agree, just to also say that explicitly somewhere.

> And further, it makes is so much more convenient to quickly add new scenes without carrying about any references to further clouds.

Yes! The ultimate goal is to make it really easy to add examples to the documentation.

> Is there maybe an option that rebuilds care only the changed files in the docs, and not everything?

Sphinx automatically only processes changed files. If there are multiple examples in one file, then the situation is more complicated... maybe we can investigate.

> And the whole build of everything could e.g. be automated to happen once in a week.

Or whenever something is deployed to master.

**Comment by leotrs** [comment]
Agreed, let's go with the sphinx directive and forget about hiding code.

As for building the documentation: like @behackl says, locally sphinx will only rebuild what needs to be rebuilt. On RTD, it will always start from scratch every time, but that's less of an issue I think.

So @behackl let's go ahead with this PR! @kolibril13 let's forget about the second repo and cloud CDN for now. Could you please close the relevant issues related to that? Thanks!

**Comment by kolibril13** [comment]
> So @behackl let's go ahead with this PR! @kolibril13 let's forget about the second repo and cloud CDN for now. Could you please close the relevant issues related to that? Thanks!

Alright.
yes, I can do that.

**Comment by behackl** [comment]
Done; the pipeline looks better than before. :-)

**Comment by behackl** [comment]
This current commit managed to deploy to https://behackl-manim.readthedocs.io/en/manim-sphinx/tutorials/quickstart.html -- basically.

The trick is that the readthedocs build image contains pretty much all of our dependencies, with the exception of `ffmpeg`. To get around that, I've prepared a conda environment (because RTD supports that) which includes ffmpeg -- and that actually works.

File: manim/utils/hashing.py [hunk 14]
```diff
@@ -41,8 +41,12 @@ def default(self, obj):
                 # NOTE : All module types objects are removed, because otherwise it throws ValueError: Circular reference detected if not. TODO
                 if isinstance(cvardict[i], ModuleType):
                     del cvardict[i]
+            try:
```
Comment (behackl) [inline manim/utils/hashing.py:44]: @huguesdevimeux This was modified in #315 and actually breaks generating the documentation here: one video for `ArrowTip` is generated from a doctest, which is read as code entered from the command line. `inspect.getsource(obj)` throws an `OSError` for this thing (I think it tries to look at a file where the code is stored, which is not possible for commands entered on the prompt -- and thus fails).

These changes I made here were just a triage to test building the docs, I am not sure whether this is a good idea -- maybe you have an opinion? :-)

File: docs/source/conf.py [hunk 15]
```diff
@@ -9,10 +9,35 @@
...
+        "binaries"
+    )
+    # the included binary is named ffmpeg-linux..., create a symlink
+    [ffmpeg_bin] = [file for file in os.listdir(ffmpeg_path) if file.startswith("ffmpeg-")]
+    os.symlink(
+        os.path.join(ffmpeg_path, ffmpeg_bin), 
+        os.path.join(ffmpeg_path, "ffmpeg")
+    )
+    os.environ["PATH"] += os.pathsep + ffmpeg_path
+
```
Comment (naveen521kk) [inline docs/source/conf.py:38]: @behackl Did this work?

Comment (behackl) [reply docs/source/conf.py:38]: No, but I am about to try something else that is more likely to work.

**Comment by behackl** [comment]
Heavens! That was a bit of a journey, sorry for all the notifications all these pushes have caused.

Anyhow: we're back to a working version, without the need of `conda` (thanks to a hint by @naveen521kk). This branch deployed successfully to https://behackl-manim.readthedocs.io/en/manim-sphinx/. There is nothing to configure on the side of RTD, everything is configured in the new `.readthedocs.yml` and the modified `docs/source/conf.py`.

Review at will! (Also, see https://github.com/ManimCommunity/manim/pull/402#discussion_r485981095 for the modification to `hashing.py`.)

**Review by leotrs — APPROVED** [review]

**Comment by leotrs** [comment]
I'm leaning toward merging this ASAP, but there is one thing that is bothering me. Every time I execute `make clean`, _all_ videos will have to be regenerated, no?

**Comment by behackl** [comment]
> I'm leaning toward merging this ASAP, but there is one thing that is bothering me. Every time I execute `make clean`, _all_ videos will have to be regenerated, no?

`make clean` clears the `build` directory; the media directory is located in `source/media`. The interesting question (can't test right now, am away from my computer) is whether caching works correctly. If it does, the building process should be very fast.

If it does not, we should think (maybe in a separate PR?) about how to enable caching for these videos, then the execution would be very fast at least. Building a separate mechanism that does that here seems, on first thought, like a re-invention of caching to me (and should thus be rather avoided). :-)
 
> Perhaps we can make it so that `make clean` removes everything _but_ those videos, and add a new `make clean-example-videos` or whatever, that removes those?

In principle, only `make cleanall` completely wipes the generated videos.

**Comment by leotrs** [comment]
Oh ok I see cool cool cool 

@ManimCommunity/scene-caching: thoughts on whether caching can work for this?

File: manim/utils/hashing.py (hunk 14 above)
Comment (huguesdevimeux) [reply manim/utils/hashing.py:44]: In fact, this issue has has been pointed out previously in #348, an we fixed it by disabling scene cachiing when the code was passed through `-`
To be very frank, i'm not a big fan of this change as it will potentially create hash collisions when function's code won't be retrieved (like, instead of raising an issue, it will just set the code to an empty string, to the letter won't have an y influence on the hash).

But I guess it's the only simple solution, so it is good to me. Could you just add a comment explaining this ? 

PS: sorry for the late reply

**Review by huguesdevimeux — CHANGES_REQUESTED** [review]

**Comment by behackl** [comment]
Actually, I just tested it locally: caching is already used correctly; the video is only regenerated if the underlying code changes.

There is just one thing that bothers me a little: I think that currently all scenes have to be named differently, or otherwise funky stuff might happen. I would rather like to take care of this in a follow-up issue, though.

**Comment by huguesdevimeux** [comment]
> Oh ok I see cool cool cool
> 
> @ManimCommunity/scene-caching: thoughts on whether caching can work for this?

No idea. Seeing the code, I don't see any reason why it should not be working, but it depends if manim is run in a temporary dir or in a constant dir. (I'm completely ignorant on Sphynx and everything)

File: manim/utils/hashing.py (hunk 14 above)
Comment (behackl) [reply manim/utils/hashing.py:44]: I agree, its not my favorite solution either; as I've said, this was more of a quick triage in order to make the docbuild process work again.

In principle, the code for generating the scene _is_ available -- after all, it is used in the `exec(...)` call! To avoid hash collision, I could hash it and pass the hash when initializing the scene... but I'd rather investigate this further in a follow-up issue.

I'll add a comment explaining what happens there (and why). Thanks for the feedback! :-)

Comment (behackl) [reply manim/utils/hashing.py:44]: 2d07edb4

**Review by leotrs — APPROVED** [review]

**Comment by kolibril13** [comment]
Note for everyone building the docs, examples can be added like this:
```
SomeTestExample
=================================

.. manim:: DotScene
   :quality: high
   :save_last_frame:

   class DotScene(Scene):
       def construct(self):
           dot = Dot().set_color(GREEN)
           self.add(dot)
           self.wait(1)


.. manim:: DotScene2
   :quality: high

   class DotScene2(Scene):
       def construct(self):
           dot = Dot().set_color(BLUE)
           self.add(dot)
           self.wait(1)
```

**Comment by behackl** [comment]
> Note for everyone building the docs, examples can be added like this:
> ```
> SomeTestExample
> =================================
> 
> .. manim:: DotScene
>    :quality: high
>    :save_last_frame:
> 
>    class DotScene(Scene):
>        def construct(self):
>            dot = Dot().set_color(GREEN)
>            self.add(dot)
>            self.wait(1)
> 
> 
> .. manim:: DotScene2
>    :quality: high
> 
>    class DotScene2(Scene):
>        def construct(self):
>            dot = Dot().set_color(BLUE)
>            self.add(dot)
>            self.wait(1)
> ```

It might even make sense to add a short remark like this to the guide for contributers!

Unfortunately, I have not found a way to render the docstring of the directive module: it would need to be somewhere in the manim module to enable that.
